flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java

```diff
@@ -130,14 +155,51 @@ public String toString() {
                     keyGroupRange, materialized.size(), nonMaterialized.size());
         }
 
-        private static Closeable asCloseable(KeyedStateHandle h) {
-            return () -> {
-                try {
-                    h.discardState();
-                } catch (Exception e) {
-                    ExceptionUtils.rethrowIOException(e);
+        private static class StreamStateHandleWrapper implements StreamStateHandle {
+            private static final long serialVersionUID = 1L;
+
+            private final KeyedStateHandle keyedStateHandle;
+
+            StreamStateHandleWrapper(KeyedStateHandle keyedStateHandle) {
+                this.keyedStateHandle = keyedStateHandle;
+            }
+
+            @Override
+            public void discardState() throws Exception {
+                keyedStateHandle.discardState();
+            }
+
+            @Override
+            public long getStateSize() {
+                return keyedStateHandle.getStateSize();
+            }
+
+            @Override
+            public FSDataInputStream openInputStream() throws IOException {
+                throw new UnsupportedOperationException("Should not call here.");
+            }
+
+            @Override
+            public Optional<byte[]> asBytesIfInMemory() {
+                throw new UnsupportedOperationException("Should not call here.");
+            }
```

Comment:
`SharedStateRegistry.registerReference` does return `StreamStateHandle` - and it can be **this** wrapper. Therefore, it's not safe to leave these methods unimplemented.
Or am I missing something?

Reply:
The returned `StreamStateHandle` would only be used in `IncrementalRemoteKeyedStateHandle`, which is safe here to just add such wrapper.

Reply:
> The returned StreamStateHandle would only be used in IncrementalRemoteKeyedStateHandle

That might change.

Reply:
Since `SharedStateRegistry` only accept `StreamStateHandle` on `registerSharedStates`, and the type of `materialized` is based on `KeyedStateHandle`, which cannot be used on register. That's why I have to introduce this wrapper. Moreover, current logic does not rely on these two interfaces.

Reply:
I understand the puprose of the wrapper. What I don't get is why it has to have these methods unimplemented. Why can't it just proxy calls to the nested handle?

Reply:
This is because the wrapped keyed state handle has the type of `KeyedStateHandle` which does not offer these two methods.

Reply:
I see, thanks for the clarification.
I think the proper way to address this would be to allow `SharedStateRegistry` to work with any type of state handles. But that's probably not something to do before the release. So maybe it makes sense to add a TODO or a ticket, WDYT?

Reply:
I think we can create a new ticket to focus this.